draft-ietf-quic-transport.md

```diff
@@ -2307,10 +2307,14 @@ signal closure.
 If the connection has been successfully established, endpoints MUST send any
 CONNECTION_CLOSE frames in a 1-RTT packet.  Prior to connection establishment a
 peer might not have 1-RTT keys, so endpoints SHOULD send CONNECTION_CLOSE frames
-in a Handshake packet.  If the endpoint does not have Handshake keys, or it is
-not certain that the peer has Handshake keys, it MAY send CONNECTION_CLOSE
-frames in an Initial packet.  If multiple packets are sent, they can be
-coalesced (see {{packet-coalesce}}) to facilitate retransmission.
+in a Handshake packet.  If the endpoint does not have Handshake keys, it SHOULD
+send CONNECTION_CLOSE frames in an Initial packet.
+
+The server may not know whether the client has Handshake keys.  In order to
```

Comment:
```suggestion
It is possible that an endpoint does not know whether the peer has Handshake keys.  In order to
```

Reply:
I'm fairly sure server to client is the only case two connection closes may need to be sent, which is why it's written with server an client.

Reply:
I have in mind 17.2.2.1 in Transport: https://tools.ietf.org/html/draft-ietf-quic-transport-20#section-17.2.2.1

"A client stops both sending and processing Initial packets when it sends its first Handshake packet."

This clearly indicates that a client will almost never be processing Initial and Handshake packets at once, so if you only send one connection close, some clients will drop it.  In all other cases with #2673 landed, the client and server should both still have Handshake keys even if they also have 1-RTT keys?

Is there something about TLS I'm implicitly assuming as well?  I can imagine I am, but I don't know what it is.

Reply:
Ahh, the two docs blur together in my mind.  This is OK.  I would suggest that you need to address both transitions here though.  The transition to Handshake and the transition to 1-RTT have the same sort of uncertainty.

Reply:
The handshake to 1-RTT transition has an overlap where the peer has both sets of keys, unlike Initial to Handshake, so I think using handshake confirmed as the signal here provides what we want?